### DISCUSSION TITLE
Latency Signing up for Saving Sessions?

### DISCUSSION THREAD

**Discussion by jamesonuk** [BottlecapDave/HomeAssistant-OctopusEnergy#627] [Q&A]
I setup the integration at about 15:00 today and I see that a saving session seemed to pop up fairly quickly afterwards.
The sensor seems to have picked up the event 
![image](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/assets/1040621/b10745fc-cb0b-4668-8748-968b222eee0e)
but the automation I created from the blueprint never fired
(I am 99.9% sure the automation was created before 15:30 but I am often wrong....)

I have signed up manually for now but I was wondering how this is processed normally? 
I am really struggling to get my head around the HA event model and work out when `octopus_energy_new_octoplus_saving_session` gets fired and the fact HA just seems to report "Detected an event" doesn't particularly help things....

**Comment by BottlecapDave** [comment]
Hello. The saving sessions are retrieved on integration startup, before the event entity has been setup. The event is only fired when new saving sessions are retrieved and it hasn't seen them before. Normally this is fine, but because you configured during a saving session it had already been retrieved and fired before your automation had been setup. Had you reloaded the integration after you setup the automation, then the event would have fired again. I'll update the logoc to not load the data upon startup.

**Comment by bazcurtis** [comment]
I have added this code as an automation and it did not run. Have I missed something

```
alias: Solar - Join Octopus Saving Session
description: Solar - Join Octopus Saving Session
trigger:
  - platform: event
    event_type: octopus_energy_new_octoplus_saving_session
condition: []
action:
  - service: octopus_energy.join_octoplus_saving_session_event
    data:
      event_code: "{{ trigger.event.data[\"event_code\"] }}"
    target:
      entity_id: event.octopus_energy_123456_octoplus_saving_session_events
  - service: persistent_notification.create
    data:
      title: Octopus Saving Sessions Updated
      message: >
        Joined new Octopus Energy saving session. It starts at {{
        trigger.event.data["event_start"].strftime('%H:%M') }} on {{
        trigger.event.data["event_start"].day }}/{{
        trigger.event.data["event_start"].month }}
  - service: notify.notify
    data:
      message: |
        Joined new Octopus Energy Saving Session
      title: Octopus Saving Sessions Updated
mode: single
```

**Reply by BottlecapDave** [reply]
Hello. When did you add it? Was there a pending saving session? Had you just configured the integration like above?

**Reply by bazcurtis** [reply]
I added it over the weekend

**Reply by Cameleon99** [reply]
@BottlecapDave how often is the integration polling the API for new saving sessions (or other data for that matter)? Just curious. 

Also a couple of sessions back I found the join service failed, I think because OE API was being battered (was hard to join via the app too), is it possible to have any error handling and retry? I did a quick google and it seems like automations can’t have any try/catch logic? Would maybe need to be in node red but then hard to test with then needing an event to test it on.

**Reply by bazcurtis** [reply]
I have to say, I have found OE notifcations about sessions really bad. I have only had one so far. I only joined yesterday due to an alert from Octo-Aid. This automation is becoming more important than it should as it is the only way I won't miss a session, I hope. It shouldn't be so "mission critical" to rely on this considering @BottlecapDave is just doing us a favour by supplying it.

**Reply by Cameleon99** [reply]
We are extremely lucky to have resourceful folk like Dave and the author of Octo-Aid to help us out. OE notifications dont work for me either, and never had. I have had to add my own notifications in HA for extra IO slots as otherwiser I wouldn't know they are happening.

**Reply by BottlecapDave** [reply]
The saving sessions are retrieved every 30 minutes, but this will increase to 15 minutes in the next update. The `new_event` event is only raised the first time an available event is retrieved for the first time. You can "force" this event to occur again if you reload the integration. However, if people are having issues with the sign up service, then the automation could probably be created in a different way (for which I'll update the blueprint). The gist is

1. Trigger on change of `event.octopus_energy_{{ACCOUNT_ID}}_octoplus_saving_session_events`
2. Condition to ensure there is at least one event (` {{ state_attr('event.octopus_energy_{{ACCOUNT_ID}}_octoplus_saving_session_events', 'available_events') | length > 0 }}`)
3. The action would then be to take the first item from the available_events array as the source of the event to join and proceed as normal.

This will mean that it will trigger whenever the saving session information is updated (which would be every 15-30 minutes depending on version under normal circumstances). It could also be manually triggered.

If that wasn't enough, you _could_ change the trigger to be on a timer, but this runs the risk of attempting to join a saving session that has already been joined.

**Reply by BottlecapDave** [reply]
I've updated the blueprint to follow the above structure, so if you are using it you can reimport the blueprint and reload the automation to see the changes.

**Reply by Cameleon99** [reply]
Probably being a bit blind but where is the blueprint? 🫣

**Reply by BottlecapDave** [reply]
https://bottlecapdave.github.io/HomeAssistant-OctopusEnergy/blueprints/#automatically-join-saving-sessions

**Reply by Cameleon99** [reply]
@BottlecapDave think there is an issue and its referencing the wrong attribute in the Blueprint. 

![image](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/assets/27414924/221c7d43-f863-451c-b0db-6d977e252cbf)


![image](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/assets/27414924/94d7b1c3-bbdd-42e3-9793-0e2eeb425d8b)


![image](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/assets/27414924/459a0f62-bba5-48c4-9392-ab258588e8c4)


Should it refer to 'code' rather than 'event_code'?

**Reply by deltamelter** [reply]
I didn't add automation from blueprint (that's new to me :wink: ) but added as per the example in docs
At first the automation failed because I had assumed incorrectly that the account_ID was taken from the installed integration.  Fixed this and ran again, only to get this error now:
```
Stopped because an error was encountered at December 14, 2023 at 11:50:24 AM (runtime: 0.01 seconds)

Error rendering data template: UndefinedError: 'dict object' has no attribute 'event'
```

```
service: octopus_energy.join_octoplus_saving_session_event
data:
  event_code: "{{ trigger.event.data[\"event_code\"] }}"
target:
  entity_id: event.octopus_energy_a_xxxxxxxx_octoplus_saving_session_events
```

**Reply by BottlecapDave** [reply]
Yup, it looks like I messed up. Teach me (which I will not learn) for pushing changes before having a chance to test them properly. I've updated the blueprint so it will at least join, but the notification might still be broken. I'll take a look at that properly when I have more time.

**Reply by Cameleon99** [reply]
No worries, there looks to be another session tomorrow (anticipation notifications have gone out), so chance for some more testing 😁 

I also got it working in Node Red, but I've no idea if you get any errors back from calling the service if it goes wrong? So doesnt achieve anything over and above the blueprint.

**Reply by deltamelter** [reply]
Thanks, joined semi-automatically now.  We will have another chance to test sometime after 14:30 when the next session for tomorrow goes up.
btw, is there a way to grab the API response from the joinevent. I see that it contains a participant number you joined at and I am curious.

**Reply by bazcurtis** [reply]
I just saw there was a Saving Session 10 minutes before hand and went to join, which I assume I wouldn't be able to. I found out that I had been signed up and the automation I posted above worked. Thanks so much @BottlecapDave.

I don't understand Blueprints. Is this a better way to do it. How do you set that up?

**Reply by BottlecapDave** [reply]
The blueprint should now be fixed regarding the notification. Please re-import the blueprint and reload your automations to see the update.

@bazcurtis Blueprints are an easy way to share automations between users. They were created because people had issues with the examples (e.g. the bits to replace were missed or replaced with the wrong entity). But they are limited in that they can't be modified (e.g. to add additional conditions or be notified in different ways). If you have an automation working, no need to change gears, just another way of doing things in HA.

**Comment by jamesonuk** [comment]
Just wanted to say there is a session today.   I was looking around on the Octopus website and it said I was signed up for a session today which I thought was cached....  Logged onto HA and saw the notification that I had been signed up.
The automation I setup from the (original) blueprint did fire this time and has worked fine 😃

**Comment by bazcurtis** [comment]
There  is another one tomorrow and it has trigged fine. Thanks again @BottlecapDave

**Reply by Cameleon99** [reply]
I got a couple of 502 errors (not the integrations fault) and it eventually managed to sign up so that’s as good as I can expect I think!

**Reply by BottlecapDave** [reply]
Yeah the new blueprint has been designed to at least attempt again when the API fails :)

**Reply by Cameleon99** [reply]
Good stuff. Appreciate the issue is with OE and not your code but OE is being consistently flakey so kudos for working with it!

**Comment by FireFury0** [comment]
From reading these comments, the *_octopus_saving_session_events  entity should fire when a new saving session is detected.  However, on my system it seems to be firing about every 15 minutes all of the time.

This doesn't actually cause problems (presumably because there aren't actually any available sessions), but it doesn't sound right?
